<a href="https://colab.research.google.com/github/Th4l3s-pr0g/estudo-IA-CC/blob/main/SimilaridadeAbtBuy.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Importar os dados das lojas em dois dataframe, uma para cada.

In [ ]:
import pandas as pd
dataFrameAbt = pd.read_csv("/content/Abt.csv")
dataFrameBuy = pd.read_csv("/content/Buy.csv")

Vamos fazer uma breve análise exploratória dos data frames.

In [ ]:
dataFrameAbt.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1081 entries, 0 to 1080
Data columns (total 4 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   id           1081 non-null   int64  
 1   name         1081 non-null   object 
 2   description  1081 non-null   object 
 3   price        418 non-null    float64
dtypes: float64(1), int64(1), object(2)
memory usage: 33.9+ KB


In [ ]:
dataFrameBuy.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1092 entries, 0 to 1091
Data columns (total 4 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   id           1092 non-null   int64  
 1   name         1092 non-null   object 
 2   description  651 non-null    object 
 3   price        590 non-null    float64
dtypes: float64(1), int64(1), object(2)
memory usage: 34.2+ KB


In [ ]:
dataFrameAbt.head()

id                                               name  \
0   0                            sony turntable pslx350h   
1   1  bose acoustimass 5 series iii speaker system a...   
2   2                               sony switcher sbv40s   
3   3                     sony 5 disc cd player cdpce375   
4   4  bose 27028 161 bookshelf pair speakers in whit...   

                                         description  price  
0  sony turntable pslx350h belt drive system 33-1...    NaN  
1  bose acoustimass 5 series iii speaker system a...  399.0  
2  sony switcher sbv40s eliminates disconnecting ...   49.0  
3  sony 5 disc cd player - cdpce375 5 disc change...    NaN  
4  bose 161 bookshelf speakers in white 161wh art...  158.0

In [ ]:
dataFrameBuy.head(20)

id                                               name  \
0    0  linksys etherfast ezxs88w ethernet switch ezxs88w   
1    1          linksys etherfast ezxs55w ethernet switch   
2    2      netgear prosafe fs105 ethernet switch fs105na   
3    3  belkin pro series high integrity vga/svga moni...   
4    4             netgear prosafe jfs516 ethernet switch   
5    5              lacie pocket floppy disk drive 706018   
6    6   canon kp 36ip print cartridge paper kit 7737a001   
7    7  kensington orbit optical trackball usb w/ps2 a...   
8    8           linksys etherfast ef4116 ethernet switch   
9    9           linksys etherfast ef4124 ethernet switch   
10  10     linksys instant gigabit eg005w ethernet switch   
11  11        linksys g wireless-g wet54g wireless bridge   
12  12  tripp lite powerverter 375-watt ultra-compact ...   
13  13      netgear prosafe gs105 ethernet switch gs105na   
14  14                     belkin monitor cable f3h982-25   
15  15  logitech driving force pro for playstation 2 9...   
16  16  belkin ac anywhere 300w power inverter f5c400-...   
17  17  linksys wireless-g range expander wireless bri...   
18  18     sony vaio notebook and ac adapter case vgpamc2   
19  19  d-link express ethernetwork high-speed docsis ...   

                                          description  price  
0   linksys etherfast 8-port 10/100 switch ( new/w...    NaN  
1                               5 x 10/100base-tx lan    NaN  
2   netgear fs105 prosafe 5 port 10/100 desktop sw...    NaN  
3                      1 x hd-15 1 x hd-15 10ft beige    NaN  
4   netgear prosafe 16 port 10/100 rackmount switc...    NaN  
5                     lacie pocket usb floppy 1.44 mb    NaN  
6                                   36 page 4 ' x 6 '   9.99  
7                                  optical usb , ps/2    NaN  
8                              16 x 10/100base-tx lan    NaN  
9                              24 x 10/100base-tx lan  64.99  
10     linksys eg005w gigabit 5-port workgroup switch    NaN  
11                           54mbps 1 x 10/100base-tx    NaN  
12  input voltage :12 v dc output voltage :120 v a...    NaN  
13      netgear prosafe 5 port gigabit desktop switch    NaN  
14                     1 x hd-15 1 x hd-15 25ft black  15.98  
15                                                NaN    NaN  
16  input voltage :12 v dc output voltage :110 v a...    NaN  
17                                             54mbps    NaN  
18  vaio neoprene notebook & ac adapter case up to...    NaN  
19                       d-link broadband cable modem    NaN

Como será feita uma análise de todos os registros de uma tabela com todos de outra, é conveniente criarmos um índice para aumentarmos a performance do treinamento.

In [ ]:
pip install recordlinkage

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 926.9/926.9 kB 14.6 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.2/1.2 MB 58.5 MB/s eta 0:00:00


In [ ]:
import recordlinkage
indexer = recordlinkage.Index()
indexer.full()

<Index>

In [ ]:
combinacoesCandidatas = indexer.index(dataFrameAbt, dataFrameBuy)

In [ ]:
combinacoesCandidatas

MultiIndex([(   0,    0),
            (   0,    1),
            (   0,    2),
            (   0,    3),
            (   0,    4),
            (   0,    5),
            (   0,    6),
            (   0,    7),
            (   0,    8),
            (   0,    9),
            ...
            (1080, 1082),
            (1080, 1083),
            (1080, 1084),
            (1080, 1085),
            (1080, 1086),
            (1080, 1087),
            (1080, 1088),
            (1080, 1089),
            (1080, 1090),
            (1080, 1091)],
           length=1180452)

Agora que o índice está criado, vamos partir para a comparação dos valores. Primeiro, vamos configurar qual abordagem será utilizada na comparação.

In [ ]:
comparador = recordlinkage.Compare()
comparador.string('name', 'name', method="jarowinkler")

<Compare>

Agora que o objeto que fará a coparação está criado, vamos comparar os valores para obter o score de similaridade.

In [ ]:
features = comparador.compute(combinacoesCandidatas, dataFrameAbt, dataFrameBuy)

In [ ]:
features

0
0    0     0.519432
     1     0.517120
     2     0.517069
     3     0.526055
     4     0.516182
...             ...
1080 1087  0.475427
     1088  0.617316
     1089  0.541204
     1090  0.497347
     1091  0.516105

[1180452 rows x 1 columns]

Vamos filtrar para exibir os produtos com mais de 0.85 de similaridade.

In [ ]:
similares = features[features[0]>0.9]
len(similares)

190

In [ ]:
similares.head(15)

0
1  154  0.964392
6  213  1.000000
18 211  0.909939
25 21   0.940841
26 201  1.000000
30 41   0.902561
32 35   0.907328
   50   0.925189
   461  0.906377
47 63   0.964706
   79   1.000000
48 63   0.964706
   79   0.988235
51 354  0.930337
57 106  0.944190

Agora que vimos que tem 190 produtos similares, vamos ver quais são

Primeiro vamos reiniciar os índices dos dataframes para deixar a consulta mais performatica

In [ ]:
dataFrameAbt = dataFrameAbt.reset_index()
dataFrameBuy = dataFrameBuy.reset_index()

Vamos avaliar os nomes e as descrições dos registros com correspondencia maior que 90%

In [ ]:
paresSimilares = []
for index_abt, index_buy in similares.index:
  record_abt = dataFrameAbt.loc[index_abt]
  record_buy = dataFrameBuy.loc[index_buy]
  paresSimilares.append((record_abt, record_buy))

Agora que foram recuperadas as correspondencias, vamos ver os registros

In [ ]:
for i, (record_abt, record_buy) in enumerate(paresSimilares):
  print("Par:",i)
  print("Registro Abt:")
  print(record_abt)
  print("Registro Buy:")
  print(record_buy)
  print("="*100)

Par: 0
Registro Abt:
index                                                          1
id                                                             1
name           bose acoustimass 5 series iii speaker system a...
description    bose acoustimass 5 series iii speaker system a...
price                                                      399.0
Name: 1, dtype: object
Registro Buy:
index                                                        154
id                                                           154
name           bose acoustimass 5 series iii speaker system 2...
description                                    2.1-channel black
price                                                      359.0
Name: 154, dtype: object
Par: 1
Registro Abt:
index                                                          6
id                                                             6
name                     kitchenaid pasta roller and cutter kpra
description    kitchenaid pasta roller and cutter k